example_queries/response_size_avg.cc

```diff
@@ -185,36 +185,36 @@ void BidiContext::onResponseHeadersInbound() {
   // From rust code, we'll pass down, a vector of vector of strings.
   // and generate following snippet for each of the inner vector.
   {
-  std::string value;
+  int64_t value;
   if (getValue({
-      "node","metadata","WORKLOAD_NAME",
```

Comment:
This looks like a bug to me.

Reply:
I think it's actually okay.  The reason it's flipping is because vertices are stored in a hashset, which have no notion of order.  But vertices aren't supposed to have any notion of order, and it winds up making the code more efficient.  Instead of checking the entire list of nodes you have to make sure you aren't adding duplicates, as you would in a vector, you just hash it again, and it's an idempotent operation.  If we changed this to a data structure where we have a notion of order, we'd have to make sure we aren't putting in duplicates, which would be a bit more inefficient.  And with a hashset, you don't wind up actually changing the correctness of the filter created;  it's easy enough to see that it's just scrambled up a little.

Reply:
The problem is that this makes it hard to keep track of changes in the compiler via reference files or any sort of diffing. In general, the compiler output should be consistent for the same input. For example, now we can not use reference files to test because the output is inconsistent. This can be solved with the use of sets that preserve insertion order, but Rust does not provide those out of the box. There is the [linked_hash_set](https://crates.io/crates/linked_hash_set/0.1.4) crate though. It might be worth trying. [Indexset](https://github.com/bluss/indexmap) might also be a good option.

Reply:
I think it was easier to implement my own serializer since both those libraries do not implement it for their hash sets, so I've got the serializer done for the vertices.  I've run it a few times in a row, and it hasn't changed how the filters come out, so I think it's fixed.

Reply:
Indexmap does implement a serializer, you just need to add: `indexmap = { version = "1.6.1", features = ["serde-1"] }
`. I think your implementation also sorts the inputs? Afaik BTreeSet does preserve insertion order but sorts the values instead.

Reply:
Index map implements a serializer, but index set does not.  We are using a set here.  What do you mean by it preserves the insertion order but sorts the values instead?  I believe it's just going through the values with the iterator object.

Reply:
Actually, I think I misinterpreted the error message.  It's not that it's not implemented, its dependency is one version of serde.  We need version 1.0 in order to use the serde feature derive (as stated here: https://serde.rs/derive.html).  So there's no version that implements them both, so it winds up giving either unimplemented error messages or version conflict error messages.

Reply:
Hmm I just tried this, it seems to work fine. Attached a patch file with the changes: 
[indexset.patch.txt](https://github.com/dyn-tracing/tracing_compiler/files/5940126/indexset.patch.txt)

Reply:
Thanks so much.  I thought that because the dependency was for serde, I should be importing them both on the serde line.  Looks like I was just using the features wrong.  That patch with the corresponding test changes should be in now.